lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,63 +1175,186 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
   return containingModule->getFiles().front() == file;
 }
 
+static bool
+doesTargetAutolinkUsingAutolinkExtract(const SwiftTargetInfo &TargetInfo,
+                                       const llvm::Triple &Triple) {
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4())
+    return true;
+
+  if (TargetInfo.OutputObjectFormat == llvm::Triple::Wasm)
+    return true;
+
+  if (Triple.isOSCygMing())
+    return true;
+
+  return false;
+}
+
+namespace {
+struct AutolinkKind {
+  enum ValueTy {
+    LLVMLinkerOptions,
+    LLVMDependentLibraries,
+    SwiftAutoLinkExtract,
+  };
+
+  ValueTy Value;
+
+  AutolinkKind(ValueTy value) : Value(value) {}
+  AutolinkKind(const AutolinkKind &kind) : Value(kind.Value) {}
+  operator enum ValueTy() { return Value; }
+
+  StringRef getSectionNameMetadata() {
+    // FIXME: This constant should be vended by LLVM somewhere.
+    switch (Value) {
+    case AutolinkKind::LLVMDependentLibraries:
+      return "llvm.dependent-libraries";
+      break;
+    case AutolinkKind::LLVMLinkerOptions:
+    case AutolinkKind::SwiftAutoLinkExtract:
+      return "llvm.linker.options";
+      break;
+    }
+  }
```

Comment:
For these types of exhaustive switches, the way that I like to do them is have all of the cases return a value and then have an llvm_unreachable at the end. Something like:

```
switch (Value) {
case X:
   return foo;
case Y:
   return bar;
}
llvm_unreachable("Exhaustive switched isn't exhaustive?!");
```